jesd204/jesd204-core.c

```diff
@@ -0,0 +1,361 @@
+/**
+ * The JESD204 framework - core logic
+ *
+ * Copyright (c) 2022 Analog Devices Inc.
+ */
+
+#include <stdlib.h>
+#include "no_os_error.h"
+#include "no_os_print_log.h"
+#include "no_os_util.h"
+#include "jesd204-priv.h"
+
+/* no-OS specific */
+static int jesd204_dev_alloc_links(struct jesd204_dev_top *jdev_top)
+{
+	struct jesd204_link_opaque *links;
+	size_t mem_size;
+	unsigned int i;
+
+	mem_size = jdev_top->num_links * sizeof(*links);
+
+	links = calloc(1, mem_size);
+	if (!links)
+		return -ENOMEM;
+	jdev_top->active_links = links;
+
+	for (i = 0; i < jdev_top->num_links; i++) {
+		links[i].jdev_top = jdev_top;
+		links[i].link_idx = i;
+		links[i].link.link_id = jdev_top->link_ids[i];
+	}
+
+	return 0;
+}
+
+/* no-OS specific */
+int jesd204_dev_register(struct jesd204_dev **jdev,
+			 const struct jesd204_dev_data *i)
+{
+	struct jesd204_dev *dev;
+
+	dev = (struct jesd204_dev *)calloc(1, sizeof(*dev));
+	if (!dev)
+		return -1;
+
+	dev->priv = (void *)calloc(1, i->sizeof_priv);
+
+	dev->dev_data = i;
+
+	*jdev = dev;
+
+	return 0;
+}
+
+/* no-OS specific */
+int jesd204_init_topology(struct jesd204_topology **topology,
+			  struct jesd204_topology_dev *devs,
+			  unsigned int devs_number)
+{
+	struct jesd204_topology *top;
+	int i, l, d = 0;
+
+	top = (struct jesd204_topology *)calloc(1, sizeof(*top));
+	if (!top)
+		return -1;
+
+	top->dev_top = (struct jesd204_dev_top *)calloc(1, sizeof(*top->dev_top));
+	if (!top->dev_top)
```

Comment:
I think we should `free(top)` here.

Reply:
yeah this looks like what cppcheck is complaining about

Reply:
Done.